### PR TITLE
provider is implemented for workspace

### PR DESCRIPTION
**Pull request by Zulalm**
A provider is implemented for the workspace page. Web and mobile workspaces is updated to have 2 arguments: workspace and workspaces. Get functions for workspace are tested and work well. Token field is added to the auth provider since it is needed for createWorkspace API call.

### CODE REVIEW

File: project/FrontEnd/collaborative_science_platform/lib/exceptions/workspace_exceptions.dart
Comment (omerfaunal) [inline project/FrontEnd/collaborative_science_platform/lib/exceptions/workspace_exceptions.dart:1]: It's really good and useful to have specific exceptions for each case. Nice work here

File: project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+  void clearAll() {
+    workspace = null;
+    workspaces = null;
+  }
+
+  Future<void> getUserWorkspaces(int id) async {
+    Uri url = Uri.parse("${Constants.apiUrl}/get_user_workspaces/?user_id=$id");
+    final Map<String, String> headers = {
+      "Accept": "application/json",
+      "content-type": "application/json"
```
Comment (omerfaunal) [inline project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart:22]: Shouldn't we add user token here? If not something wrong with backend. Workspace API should be protected with Auth (user token)

File: project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    } catch (error) {
+      rethrow;
+    }
+  }
+
+  Future<void> getWorkspaceById(int id) async {
+    Uri url = Uri.parse("${Constants.apiUrl}/get_workspace/?workspace_id=$id");
+    final Map<String, String> headers = {
+      "Accept": "application/json",
+      "content-type": "application/json"
```
Comment (omerfaunal) [inline project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart:42]: Same comment as previous

File: project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+      rethrow;
+    }
+  }
+
+  Future<void> sendCollaborationRequest(
+      int senderId, int receiverId, String title, String requestBody, int workspaceId) async {
+    Uri url = Uri.parse("${Constants.apiUrl}/send_collab_req/");
+
+    final Map<String, String> headers = {
+      "Accept": "application/json",
```
Comment (omerfaunal) [inline project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart:65]: Same comment as previous

File: project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  Future<void> editEntry(String content, int entryId) async {
+    Uri url = Uri.parse("${Constants.apiUrl}/edit_entry/");
+
+    final Map<String, String> headers = {
+      "Accept": "application/json",
```
Comment (omerfaunal) [inline project/FrontEnd/collaborative_science_platform/lib/providers/workspace_provider.dart:312]: Same comment as previous. Token issue

**Review by omerfaunal — CHANGES_REQUESTED** [review]

**Comment by omerfaunal** [comment]
My comment about these changes:

All of the endpoints in Workspace should be protected by token. User who has no permission shouldn't see or edit anything  in Workspaces that belongs to another user.

What happens if I copy endpoint and send a request not from an app, from script or something? Any Post request that changes DB or any get request shows privileged information should be protected.

fyi @ArslanArdavic @hakanaktas0

**Comment by hakanaktas0** [comment]
We plan to require session token on all post requests after the customer milestone.

**Comment by mehmetSuzer** [comment]
As @hakanaktas0 said, the session token will be implemented later. Therefore, you may skip these parts for now. After the tokens are implemented in the backend, we can revise the providers. Also, there are conflicts that needs to be resolved.

**Review by omerfaunal — APPROVED** [review]

**Comment by Zulalm** [comment]
I added session tokens to all requests related with workspaces. But they are not functioning yet.

**Pull request merged**